crates/holochain_persistence_lmdb/src/eav/lmdb.rs

```diff
@@ -0,0 +1,299 @@
+
+#[derive(Clone)]
+pub struct EavLmdbStorage<A: Attribute> {
+    id: Uuid,
+    store: SingleStore,
+    manager: Arc<RwLock<Rkv>>,
+    attribute: PhantomData<A>,
+}
+
+impl<A: Attribute> EavLmdbStorage<A> {
+    pub fn new<P: AsRef<Path> + Clone>(db_path: P) -> EavLmdbStorage<A> {
+        let cas_db_path = db_path.as_ref().join("eav").with_extension("db");
+        std::fs::create_dir_all(cas_db_path.clone())
+            .expect("Could not create file path for CAS store");
+
+        let manager = Manager::singleton()
+            .write()
+            .unwrap()
+            .get_or_create(cas_db_path.as_path(), |path: &Path| {
+                let mut env_builder = Rkv::environment_builder();
+                env_builder
+                    // max size of memory map, can be changed later
+                    .set_map_size(MAX_SIZE_BYTES)
+                    // max number of DBs in this environment
+                    .set_max_dbs(1)
+                    // Thes flags make writes waaaaay faster by async writing to disk rather than blocking
+                    // There is some loss of data integrity guarantees that comes with this
+                    .set_flags(EnvironmentFlags::WRITE_MAP | EnvironmentFlags::MAP_ASYNC);
+                Rkv::from_env(path, env_builder)
+            })
+            .expect("Could not create the environment");
+
+        let env = manager
+            .read()
+            .expect("Could not get a read lock on the manager");
+
+        // Then you can use the environment handle to get a handle to a datastore:
+        let options = StoreOptions {
+            create: true,
+            flags: DatabaseFlags::empty(),
+        };
+        let store: SingleStore = env
+            .open_single(EAV_BUCKET, options)
+            .expect("Could not create EAV store");
+
+        EavLmdbStorage {
+            id: Uuid::new_v4(),
+            store: store,
+            manager: manager.clone(),
+            attribute: PhantomData,
+        }
+    }
+}
+
+impl<A: Attribute> Debug for EavLmdbStorage<A> {
+    fn fmt(&self, f: &mut Formatter) -> Result<(), Error> {
+        f.debug_struct("EavLmdbStorage")
+            .field("id", &self.id)
+            .finish()
+    }
+}
+
+impl<A: Attribute> EavLmdbStorage<A>
+where
+    A: Sync + Send + serde::de::DeserializeOwned,
+{
+    fn add_lmdb_eavi(
+        &mut self,
+        eav: &EntityAttributeValueIndex<A>,
+    ) -> Result<Option<EntityAttributeValueIndex<A>>, StoreError> {
+        let env = self.manager.read().unwrap();
+        let mut writer = env.write()?;
+
+        // use a clever key naming scheme to speed up exact match queries on the entity
+        let key = format!("{}::{}", eav.entity(), eav.index());
+
+        self.store
+            .put(&mut writer, key, &Value::Json(&eav.content().to_string()))?;
+
+        writer.commit()?;
+
+        Ok(Some(eav.clone()))
+    }
```

Comment:
Another thing we need to keep in mind is what we should do if we find a duplicate timestamp, in all other cases if we do find a duplicate, we should generate a new EAVI and then store that and then return it

Reply:
Ah good catch. I see the PickleDB implementation was doing that as well